Feature/edit button

CurlyAnteater opened the pull request:
## Description
This pull request adds a new "Edit" button functionality to the "Rewriting Rules" tab, allowing preexisting rewriting rules to be modified.

## Changes Made
- Added a new "Edit" button to each rule in the "Rewriting Rules" tab.
- Created a frontend modal to display the "Add Rewriting Rule" ui with the rule's fields filled out and allow users to update the rewriting rule.
- Implemented the necessary backend logic and API endpoints to handle the data update process.
- Added functions on server.py, rule_manager.py data_manager.py RewritingRules.js
- Created new file EditRewritingRules.js to handle with some fields of the rule to be filled out when the "Edit" button is pressed.

<img width="1437" alt="258483746-1c02fc02-8680-4e46-a632-0d89b22b47ea" src="https://github.com/ISG-ICS/QueryBooster/assets/75426413/c9cc6b8f-4888-4ae5-adcb-98c4167fbbb0">

baiqiushi commented:
Please add screenshots for the changes to the descriptions. Thanks!

baiqiushi commented on core/data_manager.py:
```diff
@@ -195,6 +195,26 @@ def add_rule(self, rule: dict, user_id: str) -> bool:
+    def edit_rule(self, rule: dict, user_id: str, rule_id: str) -> bool:
+        try:
+            cur = self.db_conn.cursor()
+            cur.execute('''INSERT INTO rules (id, "key", name, pattern, constraints, rewrite, actions, user_id) 
```
Won't this be adding a new rule into the `rules` table instead of updating the existing rule with new values?

CurlyAnteater replied:
INSERT INTO rules (id, "key", name, pattern, constraints, rewrite, actions, user_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?) ON CONFLICT (id) DO UPDATE SET name=excluded.name, pattern=excluded.pattern, constraints=excluded.constraints, rewrite=excluded.rewrite, actions=excluded.actions
ON CONFLICT will check if there is a rule with the same id, and if there is so, DO UPDATE SET will modify the name, pattern, constraints, rewrite, and actions, leaving the id untouched.

baiqiushi commented on core/data_manager.py:
```diff
@@ -195,6 +195,26 @@ def add_rule(self, rule: dict, user_id: str) -> bool:
+                         rule['constraints'], rule['rewrite'], rule['actions'], user_id
+                         ])
+            cur.execute(
+                '''INSERT INTO internal_rules (rule_id, pattern_json, constraints_json, rewrite_json, actions_json) VALUES (?, ?, ?, ?, ?)
```
Same issue as above.

CurlyAnteater replied:
Same as above.

baiqiushi reviewed (changes requested):
See the detailed reviews.

CurlyAnteater commented:
## Changes Made
- Removed irrelevant imports and redundant parameters and code.
- Combined `AddRewritingRule.js` and `EditRewritingRule.js` to `RewritingRuleModal.js`
- `server.py` - combined `self.post_add_rule()` and `self.post_edit_rule()` to `self.post_rewrite_rule()`
- `rule_manager.py` - combined `add_rule()` and `edit_rule()` to `rewrite_rule()`
- Made so that the title and submit button of RewritingRuleModal changes depending on if a new rule is being added or if it is being edited.
- Modified few other files to adapt the changes made above.


<img width="1437" alt="Screenshot 2023-08-04 at 9 55 38 AM" src="https://github.com/ISG-ICS/QueryBooster/assets/75426413/1c02fc02-8680-4e46-a632-0d89b22b47ea">

CurlyAnteater commented on client/src/dashboard/RewritingRules.js:
```diff
@@ -113,6 +114,14 @@ export default function RewrittingRules() {
   };
 
+  const editRewritingRule = (rule) => {
+    NiceModal.show(EditRewritingRule, {user_id: user.id, rule: rule, rule_id: rule.id})
```
fixed.

CurlyAnteater commented on client/src/dashboard/RewritingRules.js:
```diff
@@ -16,6 +16,7 @@ import SyntaxHighlighter from 'react-syntax-highlighter';
 import { vs } from 'react-syntax-highlighter/dist/esm/styles/hljs';
 import { Box } from '@mui/material';
 import AddRewritingRule from './AddRewritingRule';
+import EditRewritingRule from './EditRewritingRule';
```
fixed.

CurlyAnteater commented on client/src/dashboard/EditRewritingRule.js:
```diff
@@ -0,0 +1,212 @@
+import * as React from 'react';
```
fixed.

baiqiushi commented:
> ## Changes Made
> * Removed irrelevant imports and redundant parameters and code.
> * Combined `AddRewritingRule.js` and `EditRewritingRule.js` to `RewritingRuleModal.js`
> * `server.py` - combined `self.post_add_rule()` and `self.post_edit_rule()` to `self.post_rewrite_rule()`
> * `rule_manager.py` - combined `add_rule()` and `edit_rule()` to `rewrite_rule()`
> * Improved UI so that the title and submit button of RewritingRuleModal changes depending on if a new rule is being added or if it is being edited.
> * Modified few other files to adapt the changes made above.
> 
> <img alt="Screenshot 2023-08-04 at 9 55 38 AM" width="1437" src="https://user-images.githubusercontent.com/75426413/258483746-1c02fc02-8680-4e46-a632-0d89b22b47ea.png">

I mean adding the screenshot to the original description of the PR, not adding a new comment.

CurlyAnteater commented:
> ## Changes Made
> * Removed irrelevant imports and redundant parameters and code.
> * Combined `AddRewritingRule.js` and `EditRewritingRule.js` to `RewritingRuleModal.js`
> * `server.py` - combined `self.post_add_rule()` and `self.post_edit_rule()` to `self.post_rewrite_rule()`
> * `rule_manager.py` - combined `add_rule()` and `edit_rule()` to `rewrite_rule()`
> * Improved UI so that the title and submit button of RewritingRuleModal changes depending on if a new rule is being added or if it is being edited.
> * Modified few other files to adapt the changes made above.

baiqiushi commented on core/data_manager.py:
```diff
@@ -195,6 +195,26 @@ def add_rule(self, rule: dict, user_id: str) -> bool:
             print(e)
             return False
+
+    def edit_rule(self, rule: dict, user_id: str) -> bool:
```
Since the SQL query here handles both adding a new rule and editing an existing rule, do we still need the `add_rule` function? How about unify them into one function called `save_rule`?

CurlyAnteater replied:
You are right. This is much better.

baiqiushi commented on server/server.py:
```diff
@@ -167,21 +167,25 @@ def post_disable_rule(self):
-    
-    def post_add_rule(self):
+
+    def post_rewrite_rule(self):
```
Here you already had a thought that the backend endpoint can be unified into one `post_rewrite_rule`. Then how about we unify the endpoint into one `/saveRule`? And the frontend does not have to differentiate them either, right?

CurlyAnteater replied:
I just wanted to know when debugging if a rule was coming from "add a rule" or "edit a rule". But I will change it. It seems like it won't matter too much since they share most functions.

baiqiushi commented on core/rule_manager.py:
```diff
@@ -20,13 +20,13 @@ def __init_rules(self) -> None:
         del self.dm
 
-    def add_rule(self, rule: dict, user_id: str) -> bool:
+    def rewrite_rule(self, rule: dict, user_id: str) -> bool:
```
```suggestion
    def save_rule(self, rule: dict, user_id: str) -> bool:
```

baiqiushi commented on core/data_manager.py:
```diff
@@ -176,20 +176,24 @@ def update_rule(self, rule: dict) -> None:
-    
-    def add_rule(self, rule: dict, user_id: str) -> bool:
+
+    def add_or_edit_rule(self, rule: dict, user_id: str) -> bool:
```
```suggestion
    def save_rule(self, rule: dict, user_id: str) -> bool:
```

baiqiushi commented on core/rule_manager.py:
```diff
@@ -20,13 +20,13 @@ def __init_rules(self) -> None:
         rule['actions_json'] = RuleParser.parse_actions(rule['actions'], rule['mapping'])
-        return self.dm.add_rule(rule, user_id)
-    
+        return self.dm.add_or_edit_rule(rule, user_id)
```
```suggestion
        return self.dm.save_rule(rule, user_id)
```

baiqiushi commented on server/server.py:
```diff
@@ -167,21 +167,21 @@ def post_disable_rule(self):
         rule = request['rule']
         user_id = request['user_id']
-        success = self.rm.add_rule(rule, user_id)
+        success = self.rm.rewrite_rule(rule, user_id)
```
```suggestion
        success = self.rm.save_rule(rule, user_id)
```

baiqiushi commented on server/server.py:
```diff
@@ -495,8 +495,8 @@ def do_POST(self):
-        elif self.path == "/addRule":
-            self.post_add_rule()
+        elif self.path == "/saveRule":
+            self.post_rewrite_rule()
```
```suggestion
            self.post_save_rule()
```

baiqiushi commented on server/server.py:
```diff
@@ -167,21 +167,21 @@ def post_disable_rule(self):
-    
-    def post_add_rule(self):
+
+    def post_rewrite_rule(self):
```
```suggestion
    def post_save_rule(self):
```

baiqiushi reviewed (changes requested):
I proposed to unify the backend functions with the same name, `save_rule`. Let me know if you any other thoughts. Thanks!

CurlyAnteater commented on client/src/dashboard/RewritingRules.js:
```diff
@@ -150,6 +151,9 @@ export default function RewrittingRules() {
                   <AppTagCell ruleId={rule.id} tags={rule.enabled_apps} />
                 </TableCell>
+                <TableCell align="center">
+                  <Button variant="outlined" color="error" onClick={() => AddOrEditRewritingRule(rule)} >Edit</Button>
```
This is my mistake. I will make sure to be more thorough next time.

CurlyAnteater commented:
## Changes Made
- modified function names
- new rules are pushed to the mock, edited rules are updated in the mock
- added "Add to mine" button functionality

<img width="1680" alt="Screenshot 2023-08-10 at 9 39 39 AM" src="https://github.com/ISG-ICS/QueryBooster/assets/75426413/54c10f1d-8a80-4763-9ca7-ec8d9c03835e">

baiqiushi commented on client/src/dashboard/QuerySuggestionRewritingPath.js:
```diff
@@ -53,7 +57,16 @@ const QuerySuggestionRewritingPath = NiceModal.create(({ queryId }) => {
   React.useEffect(() => {getSuggestionRewritingPath(queryId)}, []);
-  
+
+  const AddToMine = (recommendedQuery) => {
```
This part seems a bit over-complex. Each `Add To Mine` button should be correlated to a already existent rule in the rule base (although owned by another user). Then why do we need to pass the queries to the `RewritingRuleModal`? In this way, the user has to click the `formulate` button again to generate a rule, which we already have in the database. This seems not a bit confusing. I suggest leaving this PR to its original scope that does not include the `AddToMine` button changes and merge it first. Then create a new PR for the `AddToMine` button. How does that sound?

baiqiushi commented on client/src/dashboard/RewritingRuleModal.js:
```diff
@@ -17,17 +17,21 @@ import defaultRulesData from '../mock-api/listRules';
 import RuleGraph from './RuleGraph';
 
-const AddRewritingRule = NiceModal.create(({user_id}) => {
+const RewritingRuleModal = NiceModal.create(({user_id, rule=null, queries=null}) => {
```
If we remove the changes for `AddToMine`, we can remove the `queries` parameter.

baiqiushi commented on client/src/dashboard/RewritingRuleModal.js:
```diff
@@ -17,17 +17,21 @@ import defaultRulesData from '../mock-api/listRules';
   // Set up states for an example rewriting pair
-  const [q0, setQ0] = React.useState("");
-  const [q1, setQ1] = React.useState("");
+  const [q0, setQ0] = React.useState(isAddToMine ? queries[0] : "");
```
This will be reverted similarly.

baiqiushi commented on client/src/dashboard/RewritingRuleModal.js:
```diff
@@ -79,44 +83,61 @@ const AddRewritingRule = NiceModal.create(({user_id}) => {
-            "actions": "",
-            "enabled_apps": []
+
+        if(isNewRule){
```
```suggestion
        if (isNewRule) {
```

baiqiushi commented on client/src/dashboard/RewritingRuleModal.js:
```diff
@@ -79,44 +83,61 @@ const AddRewritingRule = NiceModal.create(({user_id}) => {
+              "enabled_apps": []
+            }
+          );
+        }else{
```
```suggestion
        } else {
```

baiqiushi reviewed (changes requested):
Let's split the `AddToMine` button changes to another PR and modify the PR description screenshot to use the latest version (where the `Edit` button has blue color`). Thanks!

CurlyAnteater commented:
## Changes Made
- removed "Add to mine" button feature
- cleaned code

baiqiushi commented on client/src/dashboard/QuerySuggestionRewritingPath.js:
```diff
@@ -15,6 +15,7 @@ import Title from './Title';
 import defaultSuggestionRewritingPathData from '../mock-api/suggestionRewritingPath';
 import SyntaxHighlighter from 'react-syntax-highlighter';
 import { vs } from 'react-syntax-highlighter/dist/esm/styles/hljs';
+import {userContext} from "../userContext";
```
Let's include all changes to this file to the next PR which is about `Add to Mine` button.

baiqiushi commented on client/src/dashboard/RewritingRuleModal.js:
```diff
@@ -79,44 +80,61 @@ const AddRewritingRule = NiceModal.create(({user_id}) => {
-  const onAdd = () => {
+  const onAddOrEdit = () => {
     if (pattern != "" && rewrite != "") {
       // post addRule request to server
```
```suggestion
      // post saveRule request to server
```

baiqiushi reviewed (approved):
Looks good!
